warp/verifier_backend.go

```diff
@@ -62,3 +66,70 @@ func (b *backend) verifyBlockMessage(ctx context.Context, blockHashPayload *payl
 
 	return nil
 }
+
+// verifyAddressedCall verifies the addressed call message
+func (b *backend) verifyAddressedCall(addressedCall *payload.AddressedCall) *common.AppError {
+	// Further, parse the payload to see if it is a known type.
+	parsed, err := messages.Parse(addressedCall.Payload)
+	if err != nil {
+		b.stats.IncMessageParseFail()
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: "failed to parse addressed call message: " + err.Error(),
+		}
+	}
+
+	switch p := parsed.(type) {
+	case *messages.ValidatorUptime:
+		if err := b.verifyUptimeMessage(addressedCall.SourceAddress, p); err != nil {
+			b.stats.IncUptimeValidationFail()
+			return err
+		}
+	default:
+		b.stats.IncMessageParseFail()
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("unknown message type: %T", p),
+		}
+	}
+
+	return nil
+}
+
+func (b *backend) verifyUptimeMessage(sourceAddress []byte, uptimeMsg *messages.ValidatorUptime) *common.AppError {
+	if len(sourceAddress) != 0 {
+		return &common.AppError{
+			Code:    VerifyErrCode,
+			Message: "source address should be empty for uptime message",
+		}
+	}
+
+	// first get the validator's nodeID
+	nodeID, err := b.validatorState.GetNodeID(uptimeMsg.ValidationID)
+	if err != nil {
+		return &common.AppError{
+			Code:    VerifyErrCode,
+			Message: fmt.Sprintf("failed to get validator for validationID %s: %s", uptimeMsg.ValidationID, err.Error()),
+		}
+	}
+
+	// then get the current uptime
+	currentUptime, _, err := b.uptimeCalculator.CalculateUptime(nodeID)
+	if err != nil {
+		return &common.AppError{
+			Code:    VerifyErrCode,
+			Message: fmt.Sprintf("failed to calculate uptime for nodeID %s: %s", nodeID, err.Error()),
+		}
+	}
+
+	currentUptimeSeconds := uint64(currentUptime.Seconds())
+	// verify the current uptime against the total uptime in the message
+	if currentUptimeSeconds < uptimeMsg.TotalUptime {
```

Comment:
Nit: Should this be `<=`?

Reply:
IMHO it should be rewarded if they have same uptime as requested. Also TotalUptime=0 should be rewarding everyone even without any uptime.

Reply:
Ah, yeah you're totally right I was reading this wrong. Sorry about that